File: cs/Problems/Day13.cs
namespace AOC2023.Problems;

public class Day13 : IProblem<int>
{
    public int Solve(ReadOnlySpan<char> input) => SumOfReflectionValues(input);

    private static int SumOfReflectionValues(ReadOnlySpan<char> input)
    {
        Span<Range> puzzleRanges = stackalloc Range[100];
        Span<Range> lineRanges = stackalloc Range[20];
        int puzzles = input.Split(puzzleRanges, InputReader.NewLine + InputReader.NewLine);

        int summary = 0;

        for (int i = 0; i < puzzles; i++)
        {
            var rawPuzzle = input[puzzleRanges[i]];
            int height = rawPuzzle.Split(lineRanges, InputReader.NewLine);
            int width = rawPuzzle[lineRanges[0]].Length;

            var puzzle = new char[height, width];

            for (int x = 0; x < height; x++)
            for (int y = 0; y < width; y++)
                puzzle[x, y] = rawPuzzle[lineRanges[x]][y];

            int mirrorValue = GetHorizontalMirrorValue(puzzle, height, width);

            if (mirrorValue < 0)
                mirrorValue = GetVerticalMirrorValue(puzzle, height, width);

            summary += mirrorValue;
        }

        return summary;
    }

    private static int GetHorizontalMirrorValue(char[,] input, int height, int width)
    {
        for (int i = 0; i < height - 1; i++)
        {
            for (int ji = i + 1, jd = i; ; jd--, ji++)
            {
                int reflections = 0;

                for (int x = 0; x < width; x++)
                {
                    if (input[jd, x] != input[ji, x])
                        break;

                    reflections++;
                }

                if (reflections != width)
                    break;

                if (jd == 0 || ji == height - 1)
                    return (i + 1) * 100;
            }
        }

        return -1;
    }

    private static int GetVerticalMirrorValue(char[,] input, int height, int width)
    {
        for (int i = 0; i < width - 1; i++)
        {
            for (int ji = i + 1, jd = i; ; jd--, ji++)
            {
                int reflections = 0;

                for (int y = 0; y < height; y++)
                {
                    if (input[y, ji] != input[y, jd])
                        break;

                    reflections++;
                }

                if (reflections != height)
                    break;

                if (jd == 0 || ji == width - 1)
                    return i + 1;
            }
        }

        return -1;
    }
}

File: go/problems/day13.go
package problems

import "strings"

type Day13 struct{}

func (d Day13) Solve(input string) int {
	return sumOfReflectionValues(input)
}

func sumOfReflectionValues(input string) int {
	sum := 0

	for _, rawPuzzle := range strings.Split(input, NewLine+NewLine) {
		puzzle := strings.Split(rawPuzzle, NewLine)

		value := getHorizontalMirrorValue(puzzle)
		if value < 0 {
			value = getVerticalMirrorValue(puzzle)
		}

		sum += value
	}

	return sum
}

func getHorizontalMirrorValue(puzzle []string) int {
	height := len(puzzle)
	width := len(puzzle[0])

	for i := 0; i < height-1; i++ {
		for ji, jd := i+1, i; ; jd, ji = (jd - 1), (ji + 1) {
			reflections := 0

			for x := 0; x < width; x++ {
				if puzzle[jd][x] != puzzle[ji][x] {
					break
				}

				reflections++
			}

			if reflections != width {
				break
			}

			if jd == 0 || ji == height-1 {
				return (i + 1) * 100
			}
		}
	}

	return -1
}

func getVerticalMirrorValue(puzzle []string) int {
	height := len(puzzle)
	width := len(puzzle[0])

	for i := 0; i < width-1; i++ {
		for ji, jd := i+1, i; ; jd, ji = (jd - 1), (ji + 1) {
			reflections := 0

			for y := 0; y < height; y++ {
				if puzzle[y][ji] != puzzle[y][jd] {
					break
				}

				reflections++
			}

			if reflections != height {
				break
			}

			if jd == 0 || ji == width-1 {
				return i + 1
			}
		}
	}

	return -1
}
